malaria-ios/Managers/AchievementManager.swift

```diff
@@ -0,0 +1,174 @@
+    
+    // Check for duplicate achievements.
+    for achievement in self.achievements {
+      if achievement.name == name {
+        return
+      }
+    }
+    
+    let achievement = NSEntityDescription.insertNewObjectForEntityForName("Achievement", inManagedObjectContext: context) as! Achievement
+    
+    // Check whether the required properties for this achievement exists.
+    let matchingProperties: [AchievementProperty]? =
+      properties.filter { propertyNames.contains($0.name!) }
+    
+    guard matchingProperties != nil else {
+      Logger.Warn("There are no matching properties for the achievement.")
+      return
+    }
+    
+    achievement.properties = Set(matchingProperties!)
+    achievement.name = name
+    achievement.tag = tag
+    achievement.desc = description
+    
+    CoreDataHelper.sharedInstance.saveContext(context)
+  }
+  
+  /// Gets the current value of a `AchievementProperty`.
+  
+  func getValue(propertyName: String) -> Int {
+    return properties.filter { $0.name! == propertyName }
+      .first!
+      .currentValue as! Int
+  }
+  
+  /**
+   Sets the a new value for the `AchievementProperty`.
+   It doesn't overwrite the value if the achievementProperty will be activated.
+   */
+  
+  func setValue(propertyName: String, value: Int) {
+    
+    var value = value
+    let property = properties.filter { $0.name == propertyName }.first
+    
+    // Update value only if we achieved a better (lesser or greater, depeding on the case)
+    // value than the original one, not losing the previous achieved value.
+    
+    let currentPropertyValue = property?.currentValue! as! Int
+    
+    switch property!.activation! {
+    case ">":
+      value = value > currentPropertyValue ? value : currentPropertyValue
+    case "<":
+      value = value < currentPropertyValue ? value : currentPropertyValue
+    default: return
+    }
+    
+    property!.currentValue = value;
+  }
+  
+  /// Set the current value for multiple `AchievementProperty`.
+  
+  func setValue(properties: [AchievementProperty], value: Int) {
+    for property in properties {
+      setValue(property.name!, value: getValue(property.name!) + value)
+    }
+  }
+  
+  /**
+   Returns a list of the unlocked achievements.
+   */
+  
+  func checkAchievements() -> [Achievement] {
+    var unlockedAchievements = [Achievement]();
+    
+    for achievement in achievements {
+      
+      if achievement.isUnlocked! == false {
+        var activePropertiesCount = 0
+        
+        // Check for active properties
+        let properties = achievement.properties!
+        
+        for property in properties where property.isActive() {
+          activePropertiesCount += 1
+        }
+        
+        // If all properties are active, the achievement is unlocked
+        if activePropertiesCount == achievement.properties!.count {
+          achievement.isUnlocked = true
+          unlockedAchievements.append(achievement)
+        }
```

Comment:
move all the for content to a dedicated method in the Achievement Class

achievements.filter { achivement.IsUnlocked() }